libs/full/async_mpi/include/hpx/async_mpi/mpi_future.hpp

```diff
@@ -66,6 +80,29 @@ namespace hpx { namespace mpi { namespace experimental {
             {
             }
 
+            // Used when the request was not available when constructing
+            // future_data
+            void add_callback()
+            {
+                add_request_callback(
+                        [fdp = hpx::memory::intrusive_ptr<future_data>(this)] (
+                            int status)
+                        {
+                            if (status == MPI_SUCCESS)
+                            {
+                                // mark the future as ready by setting the shared_state
+                                fdp->set_data(MPI_SUCCESS);
+                            }
+                            else
+                            {
+                                fdp->set_exception(std::make_exception_ptr(
+                                    mpi_exception(std::string("MPI function returned error code : "), status)));
```

Comment:
Would the string be better in the `mpi_exception` constructor? See related comment there.

Or even better, it would be good if the error message included something about the context. E.g. `"MPI returned error code X while handling request Y"` or something like that? Do you have access to the request through `fdp` or not now?

Reply:
Yes will move make it more verbose by calling `MPI_Errorstring` on the error code and move it in the constructor :)